File: backend/users-ms/src/main/java/com/azat4dev/demobooking/users/users_commands/domain/core/values/FirstName.java
package com.azat4dev.demobooking.users.users_commands.domain.core.values;

import com.azat4dev.demobooking.common.DomainException;
import com.azat4dev.demobooking.common.utils.Assert;
import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.io.Serializable;

@Getter
@EqualsAndHashCode
public final class FirstName implements Serializable {

    public static final int MAX_LENGTH = 255;
    private final String value;

    private FirstName(String value) {
        this.value = value;
    }

    private static void validate(String value) throws CantBeEmptyException, MaxLengthException {
        Assert.string(value, CantBeEmptyException::new).notNull().notBlank();
        Assert.string(value, MaxLengthException::new).maxLength(MAX_LENGTH);
    }

    public static FirstName checkAndMakeFromString(String value) throws CantBeEmptyException, MaxLengthException {

        Assert.notNull(value, CantBeEmptyException::new);
        final var cleanedValue = value.trim();

        validate(cleanedValue);
        return new FirstName(cleanedValue);
    }

    public static FirstName dangerMakeFromStringWithoutCheck(String value) {

        assert value != null;
        return new FirstName(value);
    }

    @Override
    public String toString() {
        return value;
    }

    public static abstract class ValidationException extends DomainException {
        public ValidationException(String message) {
            super(message);
        }
    }

    public static class CantBeEmptyException extends ValidationException {
        public CantBeEmptyException() {
            super("The first name cannot be empty");
        }

        @Override
        public String getCode() {
            return "Empty";
        }
    }

    public static class MaxLengthException extends ValidationException {
        public MaxLengthException() {
            super("The maximum length of first name is " + MAX_LENGTH);
        }

        @Override
        public String getCode() {
            return "MaxLength";
        }
    }
}

File: web-client/src/domain/auth/values/FirstName.ts
import Assert from "../../utils/Assert";
import ValidationException from "../interfaces/services/ValidationException";

class FirstName {
    public static readonly MAX_LENGTH = 50;
    public static readonly MIN_LENGTH = 1;

    private constructor(public readonly value: string) {
    }

    private static validate(value: string): void {
        Assert.notBlank(value, () => new FirstName.MinLengthException());
        Assert.minLength(
            value,
            FirstName.MIN_LENGTH,
            () => new FirstName.MinLengthException()
        );
        Assert.maxLength(
            value,
            FirstName.MAX_LENGTH,
            () => new FirstName.MaxLengthException()
        );
    }

    public static checkAndCreate = (value: string): FirstName => {
        FirstName.validate(value);
        return new FirstName(value);
    }

    public static dangerouslyCreate = (value: string): FirstName => {
        return new FirstName(value);
    }

    static readonly ValidationException = ValidationException;

    static readonly MinLengthException = class extends FirstName.ValidationException {
        public constructor() {
            super(
                `First name must be at least ${FirstName.MIN_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MinLength";
        };
    };

    static readonly MaxLengthException = class extends FirstName.ValidationException {
        public constructor() {
            super(
                `First name must be at most ${FirstName.MAX_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MaxLength";
        };
    };
}

export default FirstName;
